qiskit/algorithms/time_evolvers/classical_methods/utils.py

```diff
@@ -0,0 +1,183 @@
+
+    return observable_evolution
+
+
+def _create_obs_final(
+    ops_ev_mean: np.ndarray,
+    evolution_problem: TimeEvolutionProblem,
+) -> ListOrDict[Tuple[complex, complex]]:
+    """Creates the right output format for the final value of the auxiliary operators.
+
+    Args:
+        ops_ev_mean: Array containing the expectation value of each observable at the final timestep.
+        evolution_problem: Evolution Problem to create the output of.
+
+    Returns:
+        An output with the observables mean value at the appropiate times depending on whether
+        the auxiliary operators in the evolution problem are a `list` or a `dict`.
+
+    """
+
+    aux_ops = evolution_problem.aux_operators
+    aux_ops_evaluated = [(op_ev, 0) for op_ev in ops_ev_mean]
+    if isinstance(aux_ops, dict):
+        aux_ops_evaluated = dict(zip(aux_ops.keys(), aux_ops_evaluated))
+    return aux_ops_evaluated
+
+
+def _evaluate_aux_ops(
+    aux_ops: List[sp.csr_matrix],
+    state: np.ndarray,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Evaluates the aux operators if they are provided and stores their value.
+
+    Returns:
+        Tuple of the mean and standard deviation of the aux operators for a given state.
+    """
+    op_mean = np.array([np.real(state.conjugate().dot(op.dot(state))) for op in aux_ops])
+    return op_mean
+
+
+def _sparsify(
+    evolution_problem: TimeEvolutionProblem, steps: int, real_time: bool
+) -> Tuple[np.ndarray, List[sp.csr_matrix], sp.csr_matrix]:
+    """Returns the matrices and parameters needed for time evolution in the appropiate format.
+
+    Args:
+        evolution_problem: The definition of the evolution problem.
+        steps: number of timesteps to be performed.
+        real_time: If `True` the operators returned will correspond to the ones for real time
+            evolution, else they will correspond to imaginary time evolution.
+
+    Returns:
+        A tuple with the initial state, the list of operators to evaluate and the operator to be
+        exponentiated to perfrom one timestep.
+    """
+    # Convert the initial state and Hamiltonian into sparse matrices.
+    if isinstance(evolution_problem.initial_state, QuantumCircuit):
+        state = Statevector(evolution_problem.initial_state).data.T
+    else:
+        state = evolution_problem.initial_state.to_matrix(massive=True).transpose()
+
+    hamiltonian = evolution_problem.hamiltonian.to_spmatrix()
+
+    if isinstance(evolution_problem.aux_operators, list):
+        aux_ops = [aux_op.to_spmatrix() for aux_op in evolution_problem.aux_operators]
+    elif isinstance(evolution_problem.aux_operators, dict):
+        aux_ops = [aux_op.to_spmatrix() for aux_op in evolution_problem.aux_operators.values()]
+    else:
+        aux_ops = []
+    timestep = evolution_problem.time / steps
+    step_opeartor = -((1.0j) ** real_time) * timestep * hamiltonian
+    return (state, aux_ops, step_opeartor)
+
+
+def _evolve(
```

Comment:
This method seems to contain the main logic of the algorithm. I would see it in evolvers classes, not utils. One might think how to refactor it to minimize code duplication in those classes. Perhaps a parent class would be a way to go...